### PR TITLE
Add support for typing.Self (fix #5992)

### LINKED ISSUE #5992: Support for Python 3.11 typing.Self type
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

A new type Self was introduced in PEP 673, but now Pydantic does not support it.
This can be useful, for example, for tree structures.

The code below causes an error `RuntimeError: error checking inheritance of typing.Self (type: Self)`.

```Python
from typing import Self
from pydantic import BaseModel


class SelfRef(BaseModel):
    idx: int
    children: list[Self, ...] | None


def main():
    children = [SelfRef(idx=1, children=None), SelfRef(idx=1, children=None)]
    instance = SelfRef(idx=0, children=children)


if __name__ == '__main__':
    main()
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [X] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by Youssefares**
## Change Summary

add support for  `typing.Self` introduced in python3.11 as property annotation.

## Related issue number

fix #5992 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @adriangb

### CODE REVIEW

**Comment by Youssefares** [comment]
please review

File: tests/test_types_self.py [hunk 1]
```diff
@@ -37,4 +37,39 @@ class SelfRef(BaseModel):
...
+    """
+    Self refs should be valid in covariant direction
+    """
+
+    class SelfRef(BaseModel):
+        data: int
+        ref: Self | None = None
+
+    class SubSelfRef(SelfRef):
+        pass
```
Comment (adriangb) [inline tests/test_types_self.py:53]:
```suggestion
    class SelfRef(BaseModel):
        x: int
        ref: Self | None

    class SubSelfRef(SelfRef):
        y: str
```

Can we make this pass? If you do `SubSelfRef(x=1, ref=SelfRef(x=2), y=3)` it should fail right?

Comment (Youssefares) [reply tests/test_types_self.py]: Ha! That does fail yes. I've modified the tests and assertions. I assume this is the correct behaviour?

File: tests/test_types_self.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    params=[
+        pytest.param(typing, id='typing.Self'),
+        pytest.param(typing_extensions, id='t_e.Self'),
+    ],
+)
+def fixture_self_all(request):
+    try:
+        return request.param.Self
+    except AttributeError:
+        pytest.skip(f'Self is not available from {request.param}')
```
Comment (sydney-runkle) [inline tests/test_types_self.py:20]: Cool pattern!

File: tests/test_types_self.py (hunk 1 above)
Comment (sydney-runkle) [reply tests/test_types_self.py]: Good thought @adriangb. Yep @Youssefares, that looks like the correct behavior. Thanks for the docstrings in the tests!

File: tests/test_types_self.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    with pytest.raises(
+        ValidationError,
+        match=r'ref\.ref\s+Input should be a valid dictionary or instance of SelfRef \[type=model_type,',
+    ):
+        SelfRef(data=1, ref={'data': 2, 'ref': 3}).model_dump()
+
+
+def test_recursive_model_with_subclass(Self):
+    """
```
Comment (sydney-runkle) [inline tests/test_types_self.py:44]: Could you make these one-liner docstrings?

File: pydantic/_internal/_generate_schema.py [hunk 4]
```diff
@@ -2289,3 +2294,22 @@ def get(self) -> str | None:
...
+    def __init__(self) -> None:
+        self._stack: list[str] = []
+
+    @contextmanager
+    def push(self, type_obj: str) -> Iterator[None]:
+        self._stack.append(type_obj)
+        yield
+        self._stack.pop()
+
+    def get(self) -> str | None:
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:2311]: Could we call this `tail` to be consistent with the other stack implementations used in the generate schema logic?

Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:2311]: I see you've modeled this on `_FieldNameStack` from above, so you can disregard this - `get` is fine!

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Requested a few minor changes. Looks good overall!

@adriangb, are you ok with the stack-based implementation here? 

I haven't done a deep dive, but I feel like maybe our 4 (or 5) stacks being passed around might be able to be simplified / unified somewhat...

**Comment by adriangb** [comment]
I think we maybe need some tests that verify python inheritance behavior, not just data shape. That is, something more along the lines of your original test @Youssefares. Also some json based tests.

**Comment by sydney-runkle** [comment]
I'm good with this stack based approach for now. I do think we could revisit consolidating some of the stack based logic, but that need not fit in this PR :).

**Comment by sydney-runkle** [comment]
@Youssefares,

If you're able to add those extra test cases that Adrian mentioned by the end of the week, we can include this in the 2.7 release. Thanks a bunch for your help 🚀 this is looking great so far!

**Comment by Youssefares** [comment]
Added more tests. Also realised more work is needed for use of Self inside other types referenced as properties or used with TypeAdapter. So a slight scope expansion with a couple more places to push to the stack and a few more tests. 

Feedback appreciated!

**Review by sydney-runkle — APPROVED** [review]
@Youssefares,

Great work! Thanks for adding the additional tests for the additional use cases :). This will be a much appreciated feature in 2.7!!

**Pull request merged**